### PR TITLE
tests and handling closed events added

### PR DESCRIPTION
**Pull request by ali3-atlassian**
FEATURES COMPLETED:
- handling merged/declined/deleted events (registering webhooks with them and deleting branches when these webhooks are triggered)
- tests 
- branches are crossed out/line through branch instead of disappearing from screen

### CODE REVIEW

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:92]: what if I open a PR, then delete it, and then open a new one?

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:138]: I think this (and a few other tests here) rely on returning the internal representation of the `pullrequestStore` we cannot do that. This test can just as well do ` pullRequestStore.removePullRequest(serverId, anotherBitbucketPullRequest);`
followed by an `assert` that the old PR is still present.

**Review by dkjellin — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:208]: /nit `Optional<BitbucketServerConfiguration> serverConfiguration` or `maybeConfiguration`...

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:214]: ...and then you can shortern the variable in the lambda to `configuration` etc. Just a suggestion as I didn't immediately equate `server` with `serverConfiguration`

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookEndpoint.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookEndpoint.java:47]: There are a lot of places where you check each of the three events- MERGED, DECLINED and DELETED, despite the fact that I think Jenkins treats them all equivalently. Perhaps when the event is first marshalled, you could convert these to a generic `PULL_REQUEST_CLOSED`, so we don't need so many extra comparisons

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:187]: I'm finding this conditional a bit hard to read, as it requires referring back through the method a lot. One suggestion might be to do it like this:
`(ownedHooks.stream().noneMatch(hook -> hook.getEvents().contains(MIRROR_SYNCHRONIZED_EVENT)) && events.contains(MIRROR_SYNCHRONIZED_EVENT))`
You can do this for each conditional you have here, which removes the need for the three or four lists you have above-- but that comes at the cost of making the conditional more complex (it _definitely_ needs some comments).
Readabiliity suggestion though, if you prefer the declared lists stick with them.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java:81]: Having the various closed events in a single enum would also mean you could make this else loop explicit

**Review by mhenschke-atl — APPROVED** [review]
Looks like you have addressed most of Daniel's comments- I only have a few more that are mostly optional and don't block merging. Great work!

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketPullRequest.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketPullRequest.java:48]: You must remove the state here or you break the contract for `hashCode()` 
>If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result. 

Easiest way out here is to just remove the state from the `haschCode`
(much more around this can be found in the most excellent book Effective Java, Item 8)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStore.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStore.java]:
```suggestion
    BitbucketPullRequest getPullRequest(String key, String slug, String serverId, int pullRequestId);
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:50]: I think this javadoc (and the one below) should be moved to the interface.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java]: As mentioned during pairing, I think we should change this to return an `Optional<BitbucketPullReqest>` as that makes it adhere to the non-null return and forces callers to handle the case of not finding the PR they were looking for.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:90]: This class should be `private` the only usages I found was in the test file and they are unused at that, so I think we can make this private now.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:40]: This `cacheKey` (and every other ocarance of `CacheKey` in this class) is unused and should be removed.

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:156]: this will always be true, as you pass in the same PR store (but different variable names, but they are by reference). 
We need to update this assertion to be a bit more specific.
how about asserting that
1. `anotherBitbcketPullRequest` cannot be found (returns `Optional.empty`
2. that the PR inserted with `testAddPRWithNewKey` is still present (I _think_ that's what this test is testing?)

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:50]: we tend to not call tests from other tests. It is better that you extract out the relevant bits you want to reuse to a private method and call that instead.
(here and all other places)

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImplTest.java:195]: I think this state is wrong? the variable name indicates it should be `DELETED`

**Review by dkjellin — CHANGES_REQUESTED** [review]
I think this is starting to look good. It is mostly just small niggles left. 
A few of them really needs to be fixed (the `hasCode` one and making the `CacheKey` private springs to mind) others would be good to get fixed. 

I think some comments in the method explaining what is going on for BitbucketWebhookHandler.process (the method javadoc is really good! we just need some explanation along the way, because it is complex and hard to follow)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:118]: maybe we could do 
```
if (!Collections.disjoint(hooks, MIRROR_SYNCHRONIZED_EVENT.getEventIds())) {
```
instead? does that read better? I think it is clearer what you're trying to do? Maybe other reviewers can comment and see what they think is clearer?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java:65]: There is a check-then-act race condition here. 
I think what you need to do is essentially
```java
 ConcurrentLinkedQueue<BitbucketPullRequest> prs = pullRequests.get(cacheKey) ;
        if(prs== null || prs.isEmpty()) {
            return  Optional.empty();
        }
        return prs.stream().filter(pr -> pr.getId() == pullRequestId).findFirst();
```

**Review by dkjellin — CHANGES_REQUESTED** [review]
I think this is really close now! It is only the race condition in the PR store left to fix, and even that is only used in tests so I guess we could merge with it, but I'm going ask you to fix it before merging so we don't accidentally start to use it and have bugs later.
Looks good!

**Review by dkjellin — APPROVED** [review]
Looks good. I think there are two quite minor things to fix up. I'd be happy for you to merge this, and fix those two in a separate (smaller PR) if you prefer. Or on this one whichever you prefer most.

**Pull request merged**
